src/main/java/school/hei/haapi/service/RetakeExamService.java

```diff
@@ -76,6 +81,11 @@ private RetakeExam courseResultAndSessionToRetake(
     return retakeExam;
   }
 
+  private List<RetakeExam> getExistedRetakeExams(RetakeExamSession session, String studentId) {
+    return retakeExamRepository.findRetakeExamsBySession_IdAndStudent_Id(
```

Comment:
Why did you create a specific function for this as it is called only once and it's just a fetch from database ? Juste use the retakeExamRepository.findRetakeExamsBySession_IdAndStudent_Id method where it's needed.